src/ray/object_manager/plasma/store.cc

```diff
@@ -1059,18 +1082,97 @@ void PlasmaStore::DoAccept() {
-      // The object store is still full.
-      // NOTE(swang): There could be other requests behind this one that are
-      // actually serviceable. This may be inefficient, but eventually this
-      // request will get served and unblock the following requests, once
-      // enough objects have been spilled.
-      break;
+  // Only try to process requests if the timer is not set. If the timer is set,
+  // that means that the first request is currently not serviceable because
+  // there is not enough memory. In that case, we should wait for the timer to
+  // expire before trying any requests again.
+  if (create_timer_set_) {
+    return;
+  }
+
+  uint32_t retry_after_ms = 0;
+  bool finished = create_request_queue_.ProcessRequests(&retry_after_ms);
+  if (!finished) {
+    // Try to process requests later, after space has been made.
+    RAY_CHECK(retry_after_ms > 0);
+    auto delay = boost::posix_time::milliseconds(retry_after_ms);
+    create_timer_.reset(new boost::asio::deadline_timer(io_context_));
+    create_timer_->expires_from_now(delay);
+    create_timer_->async_wait(
+        [this](const boost::system::error_code &error) {
+          RAY_CHECK(!error);
+          create_timer_set_ = false;
+          RAY_LOG(DEBUG) << "OOM timer finished, retrying create requests";
+          ProcessCreateRequests();
+        });
+    create_timer_set_ = true;
+  }
+}
+
+void CreateRequestQueue::AddRequest(const std::shared_ptr<Client> &client, const CreateObjectCallback &request_callback) {
+  queue_.push_back({client, request_callback});
+}
+
+bool CreateRequestQueue::ProcessRequest(const std::shared_ptr<Client> &client, const CreateObjectCallback &request_callback, uint32_t *retry_after_ms) {
+  bool reply_on_oom = num_retries_ >= max_retries_;
+  bool evict_if_full = evict_if_full_;
+  if (max_retries_ == 0) {
+    // If we cannot retry, then always evict on the first attempt.
+    evict_if_full = true;
+  } else if (num_retries_ > 0) {
+    // Always try to evict after the first attempt.
+    evict_if_full = true;
+  }
+  auto status = request_callback(client, reply_on_oom, evict_if_full);
+  bool request_fulfilled = false;
+
+  if (status.IsTransientObjectStoreFull()) {
+    // The object store is full, but we should wait for space to be made
+    // through spilling, so do nothing. The caller must guarantee that
+    // ProcessRequests is called again so that we can try this request again.
+    // NOTE(swang): There could be other requests behind this one that are
+    // actually serviceable. This may be inefficient, but eventually this
+    // request will get served and unblock the following requests, once
+    // enough objects have been spilled.
+    // TODO(swang): Ask the raylet to spill enough space for multiple requests
+    // at once, instead of just the head of the queue.
+    num_retries_ = 0;
+    *retry_after_ms = delay_on_transient_oom_ms_;
+  } else if (status.IsObjectStoreFull()) {
+    num_retries_++;
```

Comment:
In this case, why don't we move the request to the end of the queue again? I guess that's the matching behavior from the previous impl, and also we can avoid starvation in this case? It seems like without doing this, we will wait without creating other queued objects until this object is actually created after all timeout?

Reply:
FIFO avoid starvation right?

Reply:
It seems like after this error occurs, we don't pop out the first entry (which causes this error)? 

```cpp
  for (auto request_it = queue_.begin();
      request_it != queue_.end(); ) {
    if (!ProcessRequest(request_it->first, request_it->second, retry_after_ms)) {
      return false;
    }
```

In this case, doesn't it keep retrying until this object is creatable?

Reply:
Yeah, all of the creation requests will be blocked until the first object gets created or fails due to OOM.